netket/operator/_hamiltonian.py

```diff
@@ -347,7 +324,108 @@ def gccf_fun(x, sections):
         return jit(nopython=True)(gccf_fun)
 
     def __repr__(self):
-        return f"Ising(J={self._J}, h={self._h}; dim={self.hilbert.size})"
+        return (
+            f"{type(self).__name__}(J={self._J}, h={self._h}; dim={self.hilbert.size})"
+        )
+
+
+@partial(jax.vmap, in_axes=(0, None, None, None))
+def _ising_mels_jax(x, edges, h, J):
+    if h == 0:
+        max_conn_size = 1
+    else:
+        max_conn_size = x.size + 1
+
+    same_spins = x[edges[:, 0]] == x[edges[:, 1]]
+    mels = jnp.empty((max_conn_size,), dtype=J.dtype)
+    mels = mels.at[0].set(J * (2 * same_spins - 1).sum())
+    if h != 0:
+        mels = mels.at[1:].set(-h)
+    return mels
+
+
+def _flip_if(cond, x, local_states):
+    # TODO here we could special-case for qubit / ising
+    # by taking -x / 1 - x
+    # i.e
+    # if local_states[0] + local_states[1] == 0:
+    #     return jnp.where(cond, -x, x)
+    # elif local_states[0] == 0:
+    #     return jnp.where(cond, local_states[1] - x, x)
+    # elif local_states[1] == 0:
+    #     return jnp.where(cond, local_states[0] - x, x)
+    # else:
+    #     ...
+    was_state_0 = x == local_states[0]
+    state_0 = jnp.asarray(local_states[0], dtype=x.dtype)
+    state_1 = jnp.asarray(local_states[1], dtype=x.dtype)
+    return jnp.where(cond ^ was_state_0, state_0, state_1)
+
+
+def _ising_conn_states_jax(x, flip, local_states):
+    max_conn_size = flip.shape[0]
+
+    x_prime = jnp.expand_dims(x, axis=1)
+    x_prime = jnp.broadcast_to(x_prime, (x.shape[0], max_conn_size, x.shape[1]))
+    x_prime = _flip_if(flip, x_prime, local_states)
+    return x_prime
+
+
+@partial(jax.jit, static_argnames=("h", "local_states"))
+def _ising_kernel_jax(x, edges, flip, h, J, local_states):
+    batch_shape = x.shape[:-1]
+    x = x.reshape((-1, x.shape[-1]))
+
+    mels = _ising_mels_jax(x, edges, h, J)
+    mels = mels.reshape(batch_shape + mels.shape[1:])
+
+    if h == 0:
+        x_prime = jnp.expand_dims(x, axis=1)
+    else:
+        x_prime = _ising_conn_states_jax(x, flip, local_states)
+    x_prime = x_prime.reshape(batch_shape + x_prime.shape[1:])
+
+    return x_prime, mels
+
+
+@jax.jit
+@partial(jax.vmap, in_axes=(0, None, None, None))
+def _ising_n_conn_jax(x, edges, h, J):
+    n_sites = x.size
+    n_conn_X = jnp.asarray(h != 0, dtype=jnp.int32) * n_sites
+    same_spins = x[edges[:, 0]] == x[edges[:, 1]]
+    # TODO duplicated with _ising_mels_jax
+    mels_ZZ = J * (2 * same_spins - 1).sum()
+    n_conn_ZZ = mels_ZZ != 0
+    return n_conn_X + n_conn_ZZ
+
+
+class IsingJax(Ising):
```

Comment:
does this implement also Ising then and not only Heisenberg?

Reply:
Yes, maybe it's better to change the title